server/src/main/java/io/crate/lucene/LuceneQueryBuilder.java

```diff
@@ -329,7 +333,18 @@ public Query visitFunction(Function function, Context context) {
 
             FunctionToQuery toQuery = functions.get(function.name());
             if (toQuery == null) {
-                return genericFunctionFilter(function, context);
+                FunctionImplementation implementation = context.nodeContext.functions().getQualified(
+                    function,
+                    context.txnCtx.sessionSettings().searchPath()
+                );
+                if (implementation instanceof Scalar<?, ?> scalar) {
+                    Query query = scalar.toQuery(function, context);
+                    if (query != null) {
+                        return query;
+                    }
+                }
+                Query query = queryFromInnerFunction(function, context);
+                return query == null ? genericFunctionFilter(function, context) : query;
```

Comment:
Couldn't we merge logic here a bit with the logic below to improve readability and complexity? 
Like defining `query` above, try to set it via `scalar.toQuery`, surround the 2nd pass below with `if (query == null)` and thus only have 1 place where `queryFromInnerFunction` +  `genericFunctionFilter` is called.

Reply:
Pushed a fixup that should improve it somewhat. `Scalar` currently doesn't implement `FunctionToQuery` - not sure if we should do that. Then the `toQuery` could be assigned to and the logic could be streamlined further.

But I'm not sure if this is worth tuning further - the whole `functions` map will be removed once the refactoring is done.